File: stripe_terminal/ios/Classes/Api/ToHostExtensions.swift
import Foundation
import StripeTerminal

extension PaymentIntentParametersApi {
    func toHost() throws -> PaymentIntentParameters {
        let b = PaymentIntentParametersBuilder(
                amount: UInt(amount),
                currency: currency
            )
            .setPaymentMethodTypes(paymentMethodTypes.map { $0.toHost() })
            .setCaptureMethod(captureMethod.toHost())
            .setMetadata(metadata)
            .setStripeDescription(description)
            .setStatementDescriptor(statementDescriptor)
            .setStatementDescriptorSuffix(statementDescriptorSuffix)
            .setReceiptEmail(receiptEmail)
            .setCustomer(customerId)
            .setApplicationFeeAmount(applicationFeeAmount?.nsNumberValue)
            .setTransferDataDestination(transferDataDestination)
            .setTransferGroup(transferGroup)
            .setOnBehalfOf(onBehalfOf)
            .setSetupFutureUsage(setupFutureUsage?.toHost())
        if let it = paymentMethodOptionsParameters { b.setPaymentMethodOptionsParameters(try it.toHost()) }
        return try b.build()
    }
}

extension TippingConfigurationApi {
    func toHost() throws -> TippingConfiguration {
        return try TippingConfigurationBuilder()
            .setEligibleAmount(eligibleAmount)
            .build()
    }
}

extension PaymentMethodOptionsParametersApi {
    func toHost() throws -> PaymentMethodOptionsParameters {
        return try PaymentMethodOptionsParametersBuilder(
            cardPresentParameters: try cardPresentParameters.toHost()
        )
            .build()
    }
}

extension CardPresentParametersApi {
    func toHost() throws -> CardPresentParameters {
        let b = CardPresentParametersBuilder()
        if let it = captureMethod { b.setCaptureMethod(it.toHost()) }
        if let it = requestedPriority { b.setRequestedPriority(it.toHost()) }
        if let it = requestExtendedAuthorization { b.setRequestExtendedAuthorization(it) }
        if let it = requestIncrementalAuthorizationSupport { b.setRequestIncrementalAuthorizationSupport(it) }
        return try b.build()
    }
}

extension CardPresentCaptureMethodApi {
    func toHost() -> CardPresentCaptureMethod {
        switch self {
        case .manualPreferred:
            return .manualPreferred
        }
    }
}

extension CardPresentRoutingApi {
    func toHost() -> CardPresentRouting {
        switch self {
        case .domestic:
            return .domestic
        case .international:
            return .international
        }
    }
}

extension PaymentMethodTypeApi {
    func toHost() -> String {
        switch (self) {
        case .cardPresent:
            return "card_present"
        case .card:
            return "card"
        case .interactPresent:
            return "interact_present"
        }
    }
}

extension CaptureMethodApi {
    func toHost() -> CaptureMethod {
        switch self {
        case .automatic:
            return .automatic
        case .manual:
            return .manual
        }
    }
}

extension PaymentIntentUsageApi {
    func toHost() -> String {
        switch self {
        case .offSession:
            return "off_session"
        case .onSession:
            return "on_session"
        }
    }
}

extension DiscoveryConfigurationApi {
    func toHost() throws -> DiscoveryConfiguration? {
        switch self {
        case let config as BluetoothDiscoveryConfigurationApi:
            return try BluetoothScanDiscoveryConfigurationBuilder()
                .setTimeout(UInt(config.timeout ?? 0))
                .setSimulated(config.isSimulated)
                .build()
        case let config as BluetoothProximityDiscoveryConfigurationApi:
            return try BluetoothProximityDiscoveryConfigurationBuilder()
                .setSimulated(config.isSimulated)
                .build()
        case _ as HandoffDiscoveryConfigurationApi:
            return nil
        case let config as InternetDiscoveryConfigurationApi:
            return try InternetDiscoveryConfigurationBuilder()
                .setSimulated(config.isSimulated)
                .setLocationId(config.locationId)
                .build()
        case let config as LocalMobileDiscoveryConfigurationApi:
            return try LocalMobileDiscoveryConfigurationBuilder()
                .setSimulated(config.isSimulated)
                .build()
        case _ as UsbDiscoveryConfigurationApi:
            return nil
        default:
            fatalError()
        }
    }
    
    func toHostDiscoveryMethod() -> DiscoveryMethod? {
        switch self {
        case _ as BluetoothDiscoveryConfigurationApi:
            return .bluetoothScan
        case _ as BluetoothProximityDiscoveryConfigurationApi:
            return .bluetoothProximity
        case _ as HandoffDiscoveryConfigurationApi:
            return nil
        case _ as InternetDiscoveryConfigurationApi:
            return .internet
        case _ as LocalMobileDiscoveryConfigurationApi:
            return .localMobile
        case _ as UsbDiscoveryConfigurationApi:
            return nil
        default:
            fatalError()
        }
    }
    
    func toHostSimulated() -> Bool {
        switch self {
        case let config as BluetoothDiscoveryConfigurationApi:
            return config.isSimulated
        case let config as BluetoothProximityDiscoveryConfigurationApi:
            return config.isSimulated
        case _ as HandoffDiscoveryConfigurationApi:
            return false
        case let config as InternetDiscoveryConfigurationApi:
            return config.isSimulated
        case let config as LocalMobileDiscoveryConfigurationApi:
            return config.isSimulated
        case _ as UsbDiscoveryConfigurationApi:
            return false
        default:
            fatalError()
        }
    }
}

extension DeviceTypeApi {
    func toHost() -> DeviceType? {
        switch self {
        case .chipper2X:
            return .chipper2X
        case .verifoneP400:
            return .verifoneP400
        case .wisePad3:
            return .wisePad3
        case .stripeM2:
            return .stripeM2
        case .wisePosE:
            return .wisePosE
        case .wisePosEDevkit:
            return .wisePosEDevKit
        case .etna:
            return .etna
        case .chipper1X:
            return .chipper1X
        case .wiseCube:
            return .wiseCube
        case .stripeS700:
            return .stripeS700
        case .stripeS700Devkit:
            return .stripeS700DevKit
        case .appleBuiltIn:
            return .appleBuiltIn
        case .cotsDevice, .wisePad3s:
            return nil
        }
    }
}


extension CartApi {
    func toHost() throws -> Cart {
        return try CartBuilder(currency: currency)
            .setTax(tax)
            .setTotal(total)
            .setLineItems(lineItems.map { try $0.toHost()} )
            .build()
    }
}

extension CartLineItemApi {
    func toHost() throws -> CartLineItem {
        return try CartLineItemBuilder(displayName: description)
            .setAmount(amount)
            .setQuantity(quantity)
            .build()
    }
}

extension SetupIntentUsageApi {
    func toHost() -> SetupIntentUsage {
        switch self {
        case .offSession:
            return .offSession
        case .onSession:
            return .onSession
        }
    }
}

File: stripe_terminal/android/src/main/kotlin/mek/stripeterminal/api/ToHostExtensions.kt
package mek.stripeterminal.api

import com.stripe.stripeterminal.external.models.*
import mek.stripeterminal.microsecondsToSeconds

fun DiscoveryConfigurationApi.toHost(): DiscoveryConfiguration? {
    return when (this) {
        is BluetoothDiscoveryConfigurationApi -> DiscoveryConfiguration.BluetoothDiscoveryConfiguration(
            isSimulated = isSimulated,
            timeout = timeout?.let { microsecondsToSeconds(it) } ?: 0,
        )

        is BluetoothProximityDiscoveryConfigurationApi -> null
        is HandoffDiscoveryConfigurationApi -> DiscoveryConfiguration.HandoffDiscoveryConfiguration()
        is InternetDiscoveryConfigurationApi -> DiscoveryConfiguration.InternetDiscoveryConfiguration(
            isSimulated = isSimulated,
            location = locationId,
        )

        is LocalMobileDiscoveryConfigurationApi -> DiscoveryConfiguration.LocalMobileDiscoveryConfiguration(
            isSimulated = isSimulated,
        )

        is UsbDiscoveryConfigurationApi -> DiscoveryConfiguration.UsbDiscoveryConfiguration(
            isSimulated = isSimulated,
            timeout = timeout?.let { microsecondsToSeconds(it) } ?: 0,
        )
    }
}

fun DeviceTypeApi.toHost(): DeviceType? {
    return when (this) {
        DeviceTypeApi.CHIPPER1_X -> DeviceType.CHIPPER_1X
        DeviceTypeApi.CHIPPER2_X -> DeviceType.CHIPPER_2X
        DeviceTypeApi.STRIPE_M2 -> DeviceType.STRIPE_M2
        DeviceTypeApi.COTS_DEVICE -> DeviceType.COTS_DEVICE
        DeviceTypeApi.VERIFONE_P400 -> DeviceType.VERIFONE_P400
        DeviceTypeApi.WISE_CUBE -> DeviceType.WISECUBE
        DeviceTypeApi.WISE_PAD3 -> DeviceType.WISEPAD_3
        DeviceTypeApi.WISE_PAD3S -> DeviceType.WISEPAD_3S
        DeviceTypeApi.WISE_POS_E -> DeviceType.WISEPOS_E
        DeviceTypeApi.WISE_POS_E_DEVKIT -> DeviceType.WISEPOS_E_DEVKIT
        DeviceTypeApi.ETNA -> DeviceType.ETNA
        DeviceTypeApi.STRIPE_S700 -> DeviceType.STRIPE_S700
        DeviceTypeApi.STRIPE_S700_DEVKIT -> DeviceType.STRIPE_S700_DEVKIT
        DeviceTypeApi.APPLE_BUILT_IN -> null
    }
}

fun CartApi.toHost(): Cart {
    return Cart.Builder(
        currency = currency,
        tax = tax,
        total = total,
        lineItems = lineItems.map { it.toHost() }
    ).build()
}

fun CartLineItemApi.toHost(): CartLineItem {
    return CartLineItem.Builder(
        description = description,
        quantity = quantity.toInt(),
        amount = amount
    ).build()
}

fun PaymentIntentUsageApi.toHost(): String {
    return when (this) {
        PaymentIntentUsageApi.OFF_SESSION -> "off_session"
        PaymentIntentUsageApi.ON_SESSION -> "on_session"
    }
}

fun PaymentIntentParametersApi.toHost(): PaymentIntentParameters {
    val b = PaymentIntentParameters.Builder(
        amount = amount,
        currency = currency,
        captureMethod = when (captureMethod) {
            CaptureMethodApi.MANUAL -> CaptureMethod.Manual
            CaptureMethodApi.AUTOMATIC -> CaptureMethod.Automatic
        },
        allowedPaymentMethodTypes = paymentMethodTypes.map {
            when (it) {
                PaymentMethodTypeApi.CARD_PRESENT -> PaymentMethodType.CARD_PRESENT
                PaymentMethodTypeApi.CARD -> PaymentMethodType.CARD
                PaymentMethodTypeApi.INTERACT_PRESENT -> PaymentMethodType.INTERAC_PRESENT
            }
        },
    )
    b.setMetadata(metadata)
    description?.let(b::setDescription)
    statementDescriptor?.let(b::setStatementDescriptor)
    statementDescriptorSuffix?.let(b::setStatementDescriptorSuffix)
    receiptEmail?.let(b::setReceiptEmail)
    customerId?.let(b::setCustomer)
    applicationFeeAmount?.let(b::setApplicationFeeAmount)
    transferDataDestination?.let(b::setTransferDataDestination)
    transferGroup?.let(b::setTransferGroup)
    onBehalfOf?.let(b::setOnBehalfOf)
    setupFutureUsage?.toHost()?.let(b::setSetupFutureUsage)
    paymentMethodOptionsParameters?.let { b.setPaymentMethodOptionsParameters(it.toHost()) }
    return b.build()
}

fun PaymentMethodOptionsParametersApi.toHost(): PaymentMethodOptionsParameters {
    return PaymentMethodOptionsParameters.Builder()
        .setCardPresentParameters(cardPresentParameters.toHost())
        .build()
}

fun CardPresentParametersApi.toHost(): CardPresentParameters {
    val b = CardPresentParameters.Builder()
    captureMethod?.let { b.setCaptureMethod(it.toHost()) }
    requestExtendedAuthorization?.let { b.setRequestExtendedAuthorization(it) }
    requestIncrementalAuthorizationSupport?.let { b.setRequestIncrementalAuthorizationSupport(it) }
    requestedPriority?.let { b.setRouting(CardPresentRoutingOptionParameters(it.toHost())) }
    return b.build()
}

fun CardPresentCaptureMethodApi.toHost(): CardPresentCaptureMethod {
    return when (this) {
        CardPresentCaptureMethodApi.MANUAL_PREFERRED -> CardPresentCaptureMethod.ManualPreferred
    }
}

fun CardPresentRoutingApi.toHost(): RoutingPriority {
    return when (this) {
        CardPresentRoutingApi.DOMESTIC -> RoutingPriority.DOMESTIC
        CardPresentRoutingApi.INTERNATIONAL -> RoutingPriority.INTERNATIONAL
    }
}

fun SetupIntentUsageApi.toHost(): String {
    return when (this) {
        SetupIntentUsageApi.ON_SESSION -> "on_session"
        SetupIntentUsageApi.OFF_SESSION -> "off_session"
    }
}